### PR TITLE
Add cluster_name variable

### PR DESCRIPTION
**Pull request by bamaralf**
Fixes #000 

Changes proposed in the pull request:

This PR adds the cluster_name variable.

### CODE REVIEW

**Comment by jvelasquez** [comment]
Hi all, this would be very useful, can it be merged or looked at?

**Comment by yupwei68** [comment]
Thanks for your contribution! 🚀  Overall, this looks almost good to me. Just please add this feature into the test in test/fixture to integrate the test CI and add this feature in example in README.

**Comment by bamaralf** [comment]
Please review the changes.

File: main.tf [hunk 1]
```diff
@@ -8,11 +8,11 @@ module "ssh-key" {
 }
 
 resource "azurerm_kubernetes_cluster" "main" {
-  name                    = "${var.prefix}-aks"
+  name                    = var.cluster_name == "" ? "${var.prefix}-aks" : var.cluster_name
   kubernetes_version      = var.kubernetes_version
   location                = data.azurerm_resource_group.main.location
   resource_group_name     = data.azurerm_resource_group.main.name
-  dns_prefix              = var.prefix
+  dns_prefix              = var.cluster_name == "" ? var.prefix : var.cluster_name
```
Comment (yupwei68) [inline main.tf:15]: Here could we keep using `var.prefix` as `dns_prefix`? The user may not want `dns_prefix` the same as the `cluster_name`?

File: main.tf [hunk 2]
```diff
@@ -124,7 +124,7 @@ resource "azurerm_kubernetes_cluster" "main" {
 
 resource "azurerm_log_analytics_workspace" "main" {
   count               = var.enable_log_analytics_workspace ? 1 : 0
-  name                = "${var.prefix}-workspace"
+  name                = var.cluster_name == "" ? "${var.prefix}-workspace" : var.cluster_name
```
Comment (yupwei68) [inline main.tf:127]: Could we add a variable as `var.cluster_workspace_name` ?

File: test/fixture/main.tf [hunk 3]
```diff
@@ -73,3 +73,13 @@ module "aks_without_monitor" {
   net_profile_pod_cidr           = "10.1.0.0/16"
   depends_on                     = [azurerm_resource_group.main]
 }
+
+module "aks_cluster_name" {
+  source                         = "../.."
+  cluster_name                   = "test-cluster"
+  resource_group_name            = azurerm_resource_group.main.name
+  enable_log_analytics_workspace = false
```
Comment (yupwei68) [inline test/fixture/main.tf:81]: Here we need to test the workspace name as well.

File: README.md [hunk 4]
```diff
@@ -37,7 +37,7 @@ module "aks" {
   client_secret                    = "your-service-principal-client-password"
   kubernetes_version               = "1.19.3"
   orchestrator_version             = "1.19.3"
-  prefix                           = "prefix"
```
Comment (yupwei68) [inline README.md:40]: Then `prefix` will still be required for `dns_prefix`

File: test/fixture/main.tf [hunk 5]
```diff
@@ -73,3 +73,13 @@ module "aks_without_monitor" {
   net_profile_pod_cidr           = "10.1.0.0/16"
   depends_on                     = [azurerm_resource_group.main]
 }
+
+module "aks_cluster_name" {
+  source                         = "../.."
+  cluster_name                   = "test-cluster"
```
Comment (yupwei68) [inline test/fixture/main.tf:79]: Then here we need a `prefix` as well.

**Review by yupwei68 — COMMENTED** [review]
Thanks for your contribution. I've left some comments for allowing possibilities for more use scenarios.

File: main.tf (hunk 1 above)
Comment (bamaralf) [reply main.tf:15]: yes, it makes sense. I am going to edit this.

File: main.tf (hunk 2 above)
Comment (bamaralf) [reply main.tf:127]: I am adding the 'cluster_analytics_workspace' variable. Do you agree with this name?

File: README.md (hunk 4 above)
Comment (bamaralf) [reply README.md:40]: Ok, I revert this change.

File: main.tf (hunk 1 above)
Comment (bamaralf) [reply main.tf:15]: `dns_prefix              = var.prefix ? var.prefix : var.cluster_name`

File: test/fixture/main.tf (hunk 3 above)
Comment (bamaralf) [reply test/fixture/main.tf]: Ok. I edited this.

File: test/fixture/main.tf (hunk 5 above)
Comment (bamaralf) [reply test/fixture/main.tf:79]: Ok. I edited this.

File: variables.tf [hunk 6]
```diff
@@ -3,9 +3,22 @@ variable "resource_group_name" {
   type        = string
 }
 
+variable "cluster_name" {
+  description = "(Optional) The name for the AKS resources created in the specified Azure Resource Group. This variable overwrites the 'prefix' var (The 'prefix' var will still be applied to the dns_prefix if it is set)"
+  type        = string
+  default     = null
+}
+
+variable "cluster_analytics_workspace" {
```
Comment (yupwei68) [inline variables.tf:12]: `cluster_log_analytics_workspace_name` would be better.

File: main.tf [hunk 7]
```diff
@@ -8,11 +8,11 @@ module "ssh-key" {
 }
 
 resource "azurerm_kubernetes_cluster" "main" {
-  name                    = "${var.prefix}-aks"
+  name                    = var.cluster_name == "" ? "${var.prefix}-aks" : var.cluster_name
   kubernetes_version      = var.kubernetes_version
   location                = data.azurerm_resource_group.main.location
   resource_group_name     = data.azurerm_resource_group.main.name
-  dns_prefix              = var.prefix
+  dns_prefix              = var.prefix ? var.prefix : var.cluster_name
```
Comment (yupwei68) [inline main.tf:15]: `dns_prefix              = var.prefix` would be fine. It's weird to define two priority differently.

File: main.tf [hunk 8]
```diff
@@ -8,11 +8,11 @@ module "ssh-key" {
 }
 
 resource "azurerm_kubernetes_cluster" "main" {
-  name                    = "${var.prefix}-aks"
+  name                    = var.cluster_name == "" ? "${var.prefix}-aks" : var.cluster_name
```
Comment (yupwei68) [inline main.tf:11]: It should be `var.cluster_name == null ? "${var.prefix}-aks" : var.cluster_name`. 
If `cluster_name` not defined, it returns error in you code.

File: variables.tf [hunk 9]
```diff
@@ -3,9 +3,22 @@ variable "resource_group_name" {
...
+  description = "(Optional) The name of the Analytics workspace"
+  type        = string
+  default     = ""
```
Comment (yupwei68) [inline variables.tf:15]: default = null

File: variables.tf [hunk 10]
```diff
@@ -3,9 +3,22 @@ variable "resource_group_name" {
...
+}
+
 variable "prefix" {
   description = "The prefix for the resources created in the specified Azure Resource Group"
   type        = string
+  default     = ""
```
Comment (yupwei68) [inline variables.tf:21]: no default value then

File: variables.tf [hunk 11]
```diff
@@ -3,9 +3,22 @@ variable "resource_group_name" {
...
+}
+
+variable "cluster_analytics_workspace" {
+  description = "(Optional) The name of the Analytics workspace"
+  type        = string
+  default     = ""
+}
+
 variable "prefix" {
   description = "The prefix for the resources created in the specified Azure Resource Group"
```
Comment (yupwei68) [inline variables.tf:19]: (Required)

File: main.tf [hunk 12]
```diff
@@ -124,7 +124,7 @@ resource "azurerm_kubernetes_cluster" "main" {
 
 resource "azurerm_log_analytics_workspace" "main" {
   count               = var.enable_log_analytics_workspace ? 1 : 0
-  name                = "${var.prefix}-workspace"
+  name                = var.cluster_analytics_workspace
```
Comment (yupwei68) [inline main.tf:127]: `name                =  var.cluster_log_analytics_workspace_name==null?"${var.prefix}-workspace": var.cluster_log_analytics_workspace_name`

**Review by yupwei68 — CHANGES_REQUESTED** [review]
Thanks again for your commits and contributions. The integration test has failed for undefined `cluster_name` and undefined `cluster_log_analytics_workspace_name`, and several grammar problems. I've left several comments.  Please take a look at them and try pass the integration test after changes. Thanks!

File: main.tf (hunk 12 above)
Comment (bamaralf) [reply main.tf]: I applied the suggested changes

File: test/fixture/main.tf [hunk 13]
```diff
@@ -73,3 +73,15 @@ module "aks_without_monitor" {
...
+  prefix                         = "prefix"
+  resource_group_name            = azurerm_resource_group.main.name
+  enable_log_analytics_workspace = true
+  cluster_analytics_workspace    = "test-cluster"
```
Comment (yupwei68) [inline test/fixture/main.tf:83]: cluster_log_analytics_workspace_name = "test-cluster"

File: variables.tf [hunk 14]
```diff
@@ -3,9 +3,22 @@ variable "resource_group_name" {
...
+  description = "(Optional) The name of the Analytics workspace"
+  type        = string
+  default     = null
+}
+
 variable "prefix" {
-  description = "The prefix for the resources created in the specified Azure Resource Group"
+  description = "(Required) The prefix for the resources created in the specified Azure Resource Group"
   type        = string
+  default     = null
```
Comment (yupwei68) [inline variables.tf:21]: Could we remove the default value for a required field?

**Review by yupwei68 — CHANGES_REQUESTED** [review]
Thanks! Some changes are not updated together. Please take a look at the comments.

File: variables.tf (hunk 14 above)
Comment (bamaralf) [reply variables.tf]: Hi @yupwei68, I removed the default value.

File: test/fixture/main.tf (hunk 13 above)
Comment (bamaralf) [reply test/fixture/main.tf]: I applied this change to the test.

**Review by yupwei68 — APPROVED** [review]
LGTM!

**Pull request merged**